src/kernel/pagecache.c

```diff
@@ -542,8 +545,18 @@ closure_function(6, 1, void, pagecache_write_sg_finish,
     status_handler completion = bound(completion);
 
     pagecache_debug("%s: pn %p, q %R, sg %p, status %v\n", __func__, pn, q, sg, s);
-    if (!is_ok(s))
+    if (!is_ok(s)) {
+        if (sg)
+            sg_list_release(sg);
+        pagecache_lock_state(pc);
+        for (int i = bound(pi); i < end; i++)  {
+            pagecache_page pp = page_lookup_nodelocked(pn, i);
+            if (pp != INVALID_ADDRESS)
+                pagecache_page_release_locked(pc, pp);
```

Comment:
Here the node is actually not locked, so the page lookup is not safe. Since all relevant pages are guaranteed to be in the node's rbtree (pagecache_write_sg inserts them if not found), we can avoid a lookup for each page, and retrieve the next page via rbnode_get_next(), as done in the loop below. Looks like the check for (!is_ok(s)) and the cleanup/release stuff can be inserted just below the existing call to page_lookup_nodelocked().

Reply:
Thanks, I fixed that.